Kamon 2.0

mladens opened the pull request:
Bumps Kamon to 2.0, introduces tracing for each internal query execution (fetches, speculative executions, retries), metrics for errors and timeouts on execution level. 

Server instrumentation removed, to be moved to separate repo

mladens committed:
Kamon 2.0 support
Trace executions

mladens committed:
Readme

mladens committed:
Metric tests

mladens committed:
Driver exceptions being swallowed fix, test

mladens committed:
Update README.md

mladens committed:
Cleanup

Remove lombok

Improve tests

ivantopo commented on build.sbt:
```diff
@@ -14,81 +14,35 @@
-val kamonTestkit        = "io.kamon"            %% "kamon-testkit"            % "1.1.3"
-val scalaExtension      = "io.kamon"            %% "kanela-scala-extension"   % "0.0.14"
+val kamonCore           = "io.kamon"            %% "kamon-core"               % "2.0.1"
+val kamonTestkit        = "io.kamon"            %% "kamon-testkit"            % "2.0.0-M4"
```
Please use the latest core, testkit and kanela versions.

ivantopo commented on kamon-cassandra-client/src/main/scala/com/datastax/driver/core/ConnectionPoolAdvice.scala:
```diff
@@ -0,0 +1,29 @@
+
+      override def onSuccess(borrowedConnection: Connection): Unit = {
+        val target = TargetResolver.getTarget(borrowedConnection.address.getAddress)
+        ClientMetrics.poolBorrow(target).record(duration(time))
```
Instead of looking up the metrics every time, why not mix in the appropriate metric group into the `HostConnectionPool` and just retrieve it from there?

ivantopo commented on kamon-cassandra-client/src/main/scala/com/datastax/driver/core/ExecutionAdvices.scala:
```diff
@@ -0,0 +1,142 @@
+
+    val executionContext = execution.context
+      .withEntry(Span.Key, executionSpan)
+      .withEntry(ParentSpanKey, clientSpan)
```
What is the logic behind storing both the client Span and the parent Span?

mladens replied:
Execution span is required for tagging and measuring execution itself. After execution is done and `fetchMore` invoked by exhausting result set iterator, we want fetch execution span to be a sibling of original one, hence carrying around the parent too.

ivantopo commented on kamon-cassandra-client/src/main/scala/com/datastax/driver/core/ExecutionAdvices.scala:
```diff
@@ -0,0 +1,142 @@
+  @Advice.OnMethodEnter
+  def onFetchMore(@Advice.This hasContext: HasContext): Scope = {
+    val clientSpan = hasContext.context.get(QueryExecutionAdvice.ParentSpanKey)
+    Kamon.storeContext(Context.of(Span.Key, clientSpan))
```
Here this shouldn't be creating a new Context that just has the Span but rather use the original Context with the new Span entry. That's in case that there is additional information in the Context that might be useful for the user.

Also, where is the Context stored by this action being cleaned up? :thinking:

ivantopo commented on kamon-cassandra-client/src/main/scala/com/datastax/driver/core/ExecutionAdvices.scala:
```diff
@@ -0,0 +1,142 @@
+
+    executionSpan.tagMetrics("dc", host.getDatacenter)
+    executionSpan.tagMetrics("rack", host.getRack)
+    executionSpan.tagMetrics("target", TargetResolver.getTarget(host.getAddress))
```
There should be an option to decide whether to tag by `target` or not because depending on the deployment size this could introduce a cardinality problem. Let's do something similar to what we do in the [common HTTP instrumentation](https://github.com/kamon-io/kamon-instrumentation-common/blob/master/src/main/resources/reference.conf#L73-L95): add a setting to decide what kind of tag it will be and by default only add the dc and rack to the span metric tags and leave target as a span tag only.

Also, please add a `cassandra` prefix to those tags.

ivantopo commented on kamon-cassandra-client/src/main/scala/com/datastax/driver/core/ExecutionAdvices.scala:
```diff
@@ -0,0 +1,142 @@
+  @Advice.OnMethodEnter
+  def onStartWriting(@Advice.This execution: HasContext): Unit = {
+    val executionSpan = execution.context.get(Span.Key)
+    executionSpan.mark("writing")
```
Please change to a more descriptive name. Something like `cassandra.connection.write-started` (or similar that makes sense).

ivantopo commented on kamon-cassandra-client/src/main/scala/com/datastax/driver/core/ExecutorQueueMetricsExtractor.scala:
```diff
@@ -0,0 +1,23 @@
+
+import kamon.instrumentation.cassandra.client.ClientMetrics.ExecutorQueueMetrics
+
+object ExecutorQueueMetricsExtractor {
```
Instead of just getting queue size, what about instrumenting the methods that create those pools on `com.datastax.driver.core.ThreadingOptions` to use the kamon-executors' instrumentation and get the entire set of metrics for them?

ivantopo commented on kamon-cassandra-client/src/main/scala/kamon/instrumentation/cassandra/client/ClientMetrics.scala:
```diff
@@ -0,0 +1,136 @@
+  def getTarget(address: InetAddress): String = address.getHostAddress
+}
+
+object ClientMetrics {
```
For all the metrics that will remain, separate the metric definition from the actual lookup. A pattern that we have been following on other modules is to define the metrics in a val, including a description of the metric (that ends up in the status page, Prometheus and so on) and then have instrument groups when more than one instrument is associated with the same logical entity. 

Besides the above, unless there is absolutely no other way to do it, please use instrument groups and mix them into the monitored entities instead of looking them up every time they are needed.

Here is an example from the Akka instrumentation that can help: https://github.com/kamon-io/kamon-akka/blob/master/instrumentation/common/src/main/scala/kamon/instrumentation/akka/AkkaMetrics.scala

ivantopo commented on kamon-cassandra-client/src/main/scala/kamon/instrumentation/cassandra/client/ClientMetrics.scala:
```diff
@@ -0,0 +1,136 @@
+    Kamon.histogram("cassandra.connection-pool.size").withTag("target", host)
+
+  def trashedConnections(host: String): Histogram =
+    Kamon.histogram("cassandra.trashed-connections").withTag("target", host)
```
Could this be a counter and incremented when HostConnectionPool trashes a connection?

ivantopo commented on kamon-cassandra-client/src/main/scala/kamon/instrumentation/cassandra/client/ClientMetrics.scala:
```diff
@@ -0,0 +1,136 @@
+      .withTag("target", host)
+
+  def connections(host: String): Histogram =
+    Kamon.histogram("cassandra.connection-pool.size").withTag("target", host)
```
Could this be a range sampler instead and capture it from the HostConnectionPool? Also.. given the super high limits on the protocol V3 it looks like the value will almost always be 1, would it be possible to have a separate metric with the total number of connections across all hosts?

mladens replied:
Moved it to HostConnectionPool, total number of connections across all host would be just this metric aggregated across all of them, or is there a specific reason for a separate metric?

ivantopo replied:
The metrics would have a different behavior. For example, if you have 10 cassandra nodes and 1 connection to each one of them then:
  - The per-host range sampler will always be recording the value 1. Aggregating the 10 individual range samplers together will still produce a value of 1. If some connections come and go, the value will always be between 0 and 1.
  - The global connections range sampler will be around 10 when all connections are open. If some connections come and go the value will always be between 0 and 10. 

Does this make the difference more clear why a separate range sampler for all connections would be useful? :thinking:

ivantopo commented on kamon-cassandra-client/src/main/scala/kamon/instrumentation/cassandra/client/ClientMetrics.scala:
```diff
@@ -0,0 +1,136 @@
+  def from(session: Session): Unit = {
+    import scala.collection.JavaConverters._
+
+    Kamon.scheduler().scheduleAtFixedRate(new Runnable {
```
I'm almost sure that we can get rid of this scheduling, but let's get back to it after all other changes are applied.

mladens committed:
Bump deps

mladens committed:
Mixin pool metrics

mladens committed:
Cleanup scope on fetchMore

mladens committed:
Prefix mark

liff commented on README.md:
```diff
@@ -20,22 +22,137 @@ Supported releases and dependencies are shown below.
+Client Span is created for every `executeAsync` invocation tagged with
+- `span.kind` - `client`
+- `cassandra.query` Query CQL
+- `cassandra.keyspace`
```
Would you perhaps be willing to consider #2 ?

mladens replied:
Done, thanks for bringing it up!

mladens committed:
Replace internal executor sampling with kamon-executors instrumentation

mladens committed:
Rename client span tags to follow opentracing semantic conventions, fixes #2

mladens committed:
Rename tags

mladens committed:
Split metric definitons from instantiation

mladens committed:
Replace scheduled hostPool sampling with advices

mladens committed:
Cleanup

mladens committed:
Use span metrics instead of queryDuration metric

mladens committed:
Mixin query metrics into a host

mladens committed:
Pool size metric update on pool init, fix tests

mladens committed:
Comments

mladens committed:
Optional rack/dc tags
Fix executions span tags

mladens committed:
Metric descriptions

ivantopo commented on kamon-cassandra-client/src/main/scala/kamon/instrumentation/cassandra/client/KamonSession.scala:
```diff
@@ -0,0 +1,150 @@
+    underlying.getState
+
+
+  def getQuery(statement: Statement): String = statement match {
```
This should also handle `com.datastax.driver.core.BatchStatement` and have a default case so it wont explode for some other unknown implementations of `Statement`.

ivantopo commented on kamon-cassandra-client/src/main/scala/kamon/instrumentation/cassandra/client/KamonSession.scala:
```diff
@@ -0,0 +1,150 @@
+  val DmlStatementPrefixes = Set("select", "insert", "update", "delete")
+}
+
+class KamonSession(underlying: Session) extends AbstractSession {
```
Please, call this something like `InstrumentedSession`.

ivantopo commented on kamon-cassandra-client/src/main/scala/kamon/instrumentation/cassandra/metrics/PoolMetrics.scala:
```diff
@@ -0,0 +1,43 @@
+  )
+}
+
+class PoolMetrics(node: TargetNode) {
```
This should be a `MetricGroup`. Following the same theme from other instrumentations, this should be be an inner class on the `PoolMetrics` object called `PoolInstruments`.

ivantopo commented on kamon-cassandra-client/src/main/scala/kamon/instrumentation/cassandra/metrics/PoolMetrics.scala:
```diff
@@ -0,0 +1,43 @@
+//TODO descriptions
+object PoolMetrics {
+  val PoolBorrowTime        = Kamon.histogram(
+    "cassandra.client.pool.borrow-time",
```
Please add the explicit parameter names here, like `name = ..... ` and `description = .... `

ivantopo commented on build.sbt:
```diff
@@ -14,81 +14,36 @@
-val kamonTestkit        = "io.kamon"            %% "kamon-testkit"            % "1.1.3"
-val scalaExtension      = "io.kamon"            %% "kanela-scala-extension"   % "0.0.14"
+val kamonCore           = "io.kamon"            %% "kamon-core"               % "2.0.2"
+val kamonTestkit        = "io.kamon"            %% "kamon-testkit"            % "2.0.0-M4"
```
The testkit version should be the same as core.

mladens replied:
`2.0.2` testkit has not been released

ivantopo commented on kamon-cassandra-client/src/main/scala/kamon/instrumentation/cassandra/Cassandra.scala:
```diff
@@ -0,0 +1,53 @@
+
+  case class TargetNode(address: String, dc: String, rack: String)
+  case class NodeTags(node: Boolean, rack: Boolean, dc: Boolean)
+  case class ClientInstrumentationConfig(samplingIntervalMillis: Long, nodeTags: NodeTags)
```
`samplingIntervalMillis` is not being used anymore, please remove it. Also, please change `Cassandra` to `CassandraInstrumentation` and `ClientInstrumentationConfig` to simple `Settings`. That's to follow the same patterns we have in other modules.

You might as well need "inline" the `NodeTags` class since it will be the only member in there so far.

ivantopo replied:
Ping on this (changing it to `Settings`).

ivantopo commented on kamon-cassandra-client/src/main/scala/kamon/instrumentation/cassandra/metrics/QueryMetrics.scala:
```diff
@@ -0,0 +1,37 @@
+import kamon.metric.Counter
+import kamon.trace.Span
+//TODO descriptions
+object QueryMetrics {
```
Same comments as in `PoolMetrics`.

ivantopo replied:
Please apply the same formatting as on PoolMetrics.

ivantopo commented on kamon-cassandra-client/src/main/scala/com/datastax/driver/core/ConnectionPoolAdvices.scala:
```diff
@@ -0,0 +1,99 @@
+
+      override def onSuccess(borrowedConnection: Connection): Unit = {
+        metrics.borrow.record(duration(time))
+        metrics.inflightPerConnection.record(borrowedConnection.inFlight.get)
```
Is it possible for `inflightPerConnection` and `inflightPerHost` to become range samplers instead of histograms?

ivantopo replied:
ping

mladens replied:
Not without a substantial revisit on the advices. This way its just exposing internal information contained within pool and connection. In order to move to range sampler we'd have to control both borrow and execution points (bring in both connection and pool info there), also make sure that we cover all exit-scenarios for a query so as not to leave sampler ever increasing. Not impossible, maybe not even hard but imho not worth it just so we can expose what is essentially same info. Please elaborate further if im missing some arguments there.

ivantopo commented on kamon-cassandra-client/src/main/scala/com/datastax/driver/core/ExecutionAdvices.scala:
```diff
@@ -0,0 +1,157 @@
+    val executionSpan = Kamon.spanBuilder(operationName).asChildOf(clientSpan).start()
+
+    metrics.tagSpanMetrics(executionSpan)
+    metrics.tagSpan(executionSpan)
```
This second call is not necessary, there is no need to repeat the metric tags as span tags as well.

ivantopo commented on kamon-cassandra-client/src/main/scala/com/datastax/driver/core/ExecutionAdvices.scala:
```diff
@@ -0,0 +1,157 @@
+    val executionSpan = execution.context.get(Span.Key)
+    executionSpan.fail(exception)
+    executionSpan.finish()
+    executionSpan.trackMetrics()
```
Why is this call here? Even if this needs to be here, any calls made after finishing a Span are noop. Also, try using the fluent APIs on spans when doing several actions on the same Span.

ivantopo commented on kamon-cassandra-client/src/main/resources/reference.conf:
```diff
@@ -6,6 +6,14 @@ kamon {
     sample-interval = 100 milliseconds
+    tracing {
+      # Tag execution metrics with target node (potentially high cardinality for large clusters)
+      tag {
```
Let's move this to follow the same conventions we have in the HTTP instrumentation: users should be able to select between `span`, `metric` and `off` for each of these tags. Take a look at [this example config](https://github.com/kamon-io/kamon-instrumentation-common/blob/master/src/main/resources/reference.conf#L68-L95).

ivantopo commented on kamon-cassandra-client/src/main/resources/reference.conf:
```diff
@@ -19,7 +27,8 @@ kanela {
       stoppable = true
       bundle-name = "DataStax Java Driver for Apache Cassandra - Core"
```
- Remove `bundle-name`
- change the name to `cassandra-client`
- remove the `show-banner` and `log-level` settings from the Kanela section

ivantopo commented on kamon-cassandra-client/src/main/resources/reference.conf:
```diff
@@ -6,6 +6,14 @@ kamon {
   cassandra {
     # Interval at which all registered cassandra metrics will be sampled.
     sample-interval = 100 milliseconds
```
Remove this setting since it is no longer used.

ivantopo commented on kamon-cassandra-client/src/main/scala/kamon/instrumentation/cassandra/executors/ExecutorInstrumentation.scala:
```diff
@@ -0,0 +1,62 @@
+trait ExecutorMetrics {
+
+  def metricName(executorName: String) = "cassandra.client.executor." + executorName
+  val componentTags = TagSet.of("component", "cassandra-client")
```
Call it `cassandra.client`

mladens committed:
Rename session, handle BatchStatement

mladens committed:
Use InstrumentGroups

mladens committed:
Address more comments..

mladens committed:
Comments

mladens committed:
Comments

mladens committed:
Comments..

mladens committed:
More comments

mladens committed:
Travis

mladens committed:
Testkit snapshot

mladens committed:
Make Travis happy again

mladens committed:
Prefix tags

mladens committed:
Add global connection count across all targets

mladens committed:
Cleanup

mladens committed:
Renames

mladens committed:
Bump kanela

ivantopo commented:
Hey @mladens, I was testing this in tiny app and found a few issues by examining the data on both the Status Page and some of the data that came to Kamon APM. Here is the summary of my findings:

- [ ]  We should add the "cluster" tag to everything related to the same Cluster. The cluster name setting can be added when creating a Client Cluster although it does not have anything to do with the actual cassandra cluster name. It is specially important if a single app is connecting to more than one Cassandra cluster
- [ ]  Missing instrumentation description for the Status Page. Just include a description on the `reference.conf` file.
- [ ]  These metrics do not have any tags:
    - [ ]  cassandra.query.speculative
    - [ ]  cassandra.query.retries
- [ ]  These metrics have an "unknown" value on the `cassandra.dc` tag:
    - [ ]  cassandra.query.errors
    - [ ]  cassandra.query.timeouts
- [ ]  The cassandra.client.pool.size metric has two instances, one without any tags and one with the cassandra.dc tag, why is that?
- [ ]  Should we actually have a cassandra.query.errors metric? isn't it the same as looking at the aggregate of all span.processing-time from cassandra with error=true? I sort of see how this can be a quick go to metric for cassandra-related errors. In that case, what tags will be applied to this tag? 
- [ ]  cassandra.client.inflight-per-target should probably be called cassandra.client.pool.target.in-flight.
- [ ]  cassandra.client.pool.inflight-per-connection should probably be cassandra.client.pool.connection.in-flight.
- [ ]  The cassandra.[dc|rack|node] are missing from Spans. This is weird because I already saw these tags in our traces with a previous version of the instrumentation!
- [ ]  It would be nice if all implementation Spans (like retries and executions) could be turned on or off via config. This is a nice to have thing, if there is no clear/quick way of implementing it right now lets create an issue and tackle this on the next version.

mladens committed:
conf description

mladens committed:
Fix Host location info initialization
Rename certain metric
Cleanups

mladens committed:
Cluster tag
Update readme

mladens commented:
Comments not addressed are fixed...

> * [ ]   These metrics do not have any tags:
>   
>   * [ ]   cassandra.query.speculative
>   * [ ]   cassandra.query.retries

Those dont have any sensible dimensions to them,   
     - host that triggered retry - this is already covered by latency and error metrics
     - host that it got retried on, has nothing to do with him
     - similar line of reasoning goes for speculatives
 

> * [ ]   The cassandra.client.pool.size metric has two instances, one without any tags and one with the cassandra.dc tag, why is that?

Metrics are created per target host (depending on config) and one without tags to represent global count across all entire cluster (as requested in one of previous comments)

> * [ ]   Should we actually have a cassandra.query.errors metric? isn't it the same as looking at the aggregate of all span.processing-time from cassandra with error=true? I sort of see how this can be a quick go to metric for cassandra-related errors. In that case, what tags will be applied to this tag?

Could go both ways, I agree that it is a bit redundant. Not sure what is a kamon-wide policy here but i opted for having extra metrics on account of having metrics more understandable and easier to discover/document. Am open to revisiting this decision.


> * [ ]   It would be nice if all implementation Spans (like retries and executions) could be turned on or off via config. This is a nice to have thing, if there is no clear/quick way of implementing it right now lets create an issue and tackle this on the next version.

Not a quick fix im afraid :/ Problem is mostly in transfering execution tags (target..etc) from per-execution span into one client span that could cover multiple executions (multiple hosts, retries, fetches...etc)

ivantopo commented on kamon-cassandra-client/src/main/scala/kamon/instrumentation/cassandra/metrics/PoolMetrics.scala:
```diff
@@ -0,0 +1,46 @@
+import kamon.tag.TagSet
+
+object PoolMetrics {
+  val PoolBorrowTime        = Kamon.histogram(
```
This one should be a timer.

ivantopo commented on kamon-cassandra-client/src/main/scala/kamon/instrumentation/cassandra/metrics/PoolMetrics.scala:
```diff
@@ -0,0 +1,46 @@
+  val ConnectionPoolGlobalSize    = Kamon.rangeSampler(
+    name = "cassandra.client.pool.size",
+    description = "Total number of connections across all target nodes"
+  )
```
These two vals are effectively pointing to the same metric which ends up producing this warning when the application starts:

```
WARN  kamon.metric.MetricRegistry - Ignoring new description [Total number of connections across all target nodes] for metric [cassandra.client.pool.size]
```

The purpose of these two metrics is different so they should be kept as different metrics. Since the pool is actually more like a pool of pools we should make a better distinction between what is meant to track the behavior of the general pool and what is meant to track the behavior of the connection pool to a specific target (which, as far as I can understand will almost always be just 1 connection per target).

Based on the above, I would suggest to use:
  - `cassandra.client.pool.size` for the general pool size, meaning that it counts all connections regardless of the target.
  - `cassandra.client.pool.target.pool.size` (not sold on that name, open to suggestions) for the pools for each one of the targets. For this metric to actually be of value it should be possible to tag it with the target, otherwise it might not be very useful at all.

mladens replied:
Leaving those two metrics there was a mistake, since both of them are measuring identical thing with or without the target dimension, i'd much prefer to keep one metric (albeit with a more generic description) and have 2 separate instances on the `PoolInstruments` object.

ivantopo commented on kamon-cassandra-client/src/main/scala/kamon/instrumentation/cassandra/metrics/PoolMetrics.scala:
```diff
@@ -0,0 +1,46 @@
+  val InFlightPerTarget     = Kamon.histogram(
+    name = "cassandra.client.pool.target.in-flight",
+    description = "Number of in-flight request towards this host measured at the moment a new query is issued"
+  )
```
These metrics have a similar situation as the above: we need to differentiate whether they belong to the general pool or the per-target pool, and the per-target values should be tagged with the target in order to be actually useful. This entire per-target tagging can be deadly because of how many time series it could generate. Would it be possible to keep the per-target stuff behind a config setting?

mladens replied:
This is a similar situation only if we assume 1 connection per target host which is not neccessarily always true. Idea was to draw conclusions such as
- how saturated is the connection, should we increase pool size for (`connection.in-flight`)
- is there any disbalance in our query patterns, partitioning might be skewed (`target.in-flight`)

ivantopo commented on kamon-cassandra-client/src/main/scala/kamon/instrumentation/cassandra/metrics/QueryMetrics.scala:
```diff
@@ -0,0 +1,36 @@
+    * one defined by a policy so we are dropping it altogether */
+    val retries: Counter                = RetriedExecutions.withoutTags()
+    val speculative: Counter            = SpeculativeExecutions.withoutTags()
+    val cancelled: Counter              = CanceledExecutions.withoutTags()
```
These metrics should have the same tags as errors/timeouts. So far it seems like only `cassandra.dc` and `cassandra.cluster` will make it here. Doesn't say a lot but at least will keep all metrics consistent within the same instrument group.

ivantopo reviewed (changes requested):
Hey @mladens, I replied to your comments on the code itself so we can keep independent discussions on each of the things that need work.

Regarding other general stuff:
  - Keeping `cassandra.query.errors`. Yes, we keep it.
  - Toggling implementation spans? Future version.
  - Sorry about the corrections on the connection pool metrics. I'm still reasoning about the whole pool of pools thing and want to make sure we will nail this thing.

mladens committed:
Align instrument types

mladens committed:
Formatting

mladens committed:
Tag speculatives, retries and canceled executions with target tags

mladens committed:
Readme update

mladens committed:
Move speculative to tag instead of operation, rename retry and speculative tags

mladens committed:
Remove speculation operation

mladens committed:
Added Session metrics

mladens committed:
Added query counters to pool instruments

mladens committed:
Introduce MetricProxy routing to global metrics and host metrics if enabled
Removed QueryMetrics, move all execution metrics to Session/Pool metrics
Added HostPoolMetrics config key

mladens committed:
Rename conf target to host

mladens committed:
Session inFlight metric

mladens committed:
Periodically sample HostConnectionPool for internal in-flight count
Store registration and clear on pool close

mladens committed:
Moving around, todos...

mladens committed:
PERFORMANCE OPTIMIZATION

mladens committed:
Borrow replace timer with tdiff

mladens committed:
Tag dictionary

mladens committed:
All tags on pool metrics, removed reconfiguration capabilities

mladens committed:
Tag error metric with client/serve source tag

mladens committed:
Removed readme

mladens committed:
Clean

mladens committed:
Cross versions

mladens committed:
Scalafmt conf

mladens committed:
Reformat

mladens committed:
Fix test

mladens committed:
Fmt maxColumn 120

mladens committed:
Fix metric prefix

mladens committed:
Another fix

mladens committed:
Move all to root project

mladens committed:
Hardcode keyspace for each spec

ivantopo committed:
cosmetic changes around the codebase before merging

dpsoft reviewed (approved):
@mladens awesome!!

dispalt commented:
I'd love to use this, nice work.

mladens commented:
Kamon got moved to monorepo, closing this one, new PR here: https://github.com/kamon-io/Kamon/pull/763
